chain/network/src/routing/routing_table_state.rs

```diff
@@ -0,0 +1,404 @@
+    fn fetch_edges_for_peer_from_disk(&mut self, other_peer_id: &PeerId) {
+        if other_peer_id == self.my_peer_id()
+            || self.peer_last_time_reachable.contains_key(other_peer_id)
+        {
+            return;
+        }
+
+        let my_peer_id = self.my_peer_id().clone();
+
+        // Get the "row" (a.k.a nonce) at which we've stored a given peer in the past (when we pruned it).
+        if let Ok(component_nonce) = self.component_nonce_from_peer(other_peer_id) {
+            let mut update = self.store.store_update();
+
+            // Load all edges that were persisted in database in the cell - and add them to the current graph.
+            if let Ok(edges) = self.get_and_remove_component_edges(component_nonce, &mut update) {
+                for edge in edges {
+                    for &peer_id in vec![&edge.key().0, &edge.key().1].iter() {
+                        if peer_id == &my_peer_id
+                            || self.peer_last_time_reachable.contains_key(peer_id)
+                        {
+                            continue;
+                        }
+
+                        // `edge = (peer_id, other_peer_id)` belongs to component that we loaded from database.
+                        if let Ok(cur_nonce) = self.component_nonce_from_peer(peer_id) {
+                            // If `peer_id` belongs to current component
+                            if cur_nonce == component_nonce {
+                                // Mark it as reachable and delete from database.
+                                self.peer_last_time_reachable
+                                    .insert(peer_id.clone(), Instant::now() - SAVE_PEERS_MAX_TIME);
+                                update.delete(
+                                    DBCol::PeerComponent,
+                                    peer_id.try_to_vec().unwrap().as_ref(),
+                                );
+                            }
+                        }
+                    }
+                    self.add_verified_edge(edge);
+                }
+            }
+
+            if let Err(e) = update.commit() {
+                warn!(target: "network", "Error removing network component from store. {:?}", e);
+            }
+        } else {
+            self.peer_last_time_reachable.insert(other_peer_id.clone(), Instant::now());
+        }
+    }
+
+    pub(super) fn my_peer_id(&self) -> &PeerId {
+        self.raw_graph.my_peer_id()
+    }
+
+    /// Recalculate routing table and update list of reachable peers.
+    pub fn recalculate_routing_table(&mut self) {
+        let _d = delay_detector::DelayDetector::new(|| "routing table update".into());
+        let _routing_table_recalculation =
+            metrics::ROUTING_TABLE_RECALCULATION_HISTOGRAM.start_timer();
+
+        trace!(target: "network", "Update routing table.");
+
+        self.peer_forwarding = Arc::new(self.raw_graph.calculate_distance());
```

Comment:
this and prune-edges are two potentially long operations.

And currently they will be done under the lock - so they can also cause peer_manager_actor to get 'stuck' waiting for this lock to be released.

Reply:
Good catch, thank you. I'll move this change to a separate PR, because to address this issue a major refactor is required.